### ISSUE TITLE
Same wall object but different sematic segmentation mask color 

### ISSUE THREAD

**Issue by phongnhhn92** [allenai/ai2thor#183]
Hi guys, I am trying to perform semantic segmentation using data generated from AI2Thor for my research. However, I can not find in the docs that how many classes do AI2Thor have for the semantic maps along with their code color. As you can see, I have 2 pairs of images (rgb and class_segmentation_frame) from Scene 1 and 15. But it is clear to see that the walls in those seg frame are in different color. Is that a bug or something you guys intend to do ? In the doc, u guys mentioned that there are 105 object types within AI2-THOR but I guess the number of object classes is a larger number than 105, is that correct.
RGB scene 1
![img1](https://user-images.githubusercontent.com/11288381/57573070-8c6ed400-742b-11e9-81db-9ecefc445d67.jpg)
RGB scene 15
![img15](https://user-images.githubusercontent.com/11288381/57573073-9395e200-742b-11e9-8bab-06bc19b9bc78.jpg)
Seg_frame 1
![seg1](https://user-images.githubusercontent.com/11288381/57573077-9e507700-742b-11e9-9bca-bac5a78a94e8.jpg)
Seg_frame 15
![seg15](https://user-images.githubusercontent.com/11288381/57573079-a3adc180-742b-11e9-9ca3-487c355b8bb7.jpg)

Another issue I found is that, in the new version there is a visible blue agent in the scene moving around, is there anyway to hide it so that I can manage to gather sampled views without the agent and its shadow or reflection ?

**Comment by roozbehm** [comment]
There is no guarantee that you see the same color across scenes. You should use `color_to_object_id` in meta-data to find the mapping from the color to the string.

**Comment by phongnhhn92** [comment]
@roozbehm 
I understand your point but I just feel weird. It is really non-intuitive to have many color codes for a single "Wall" class. I can look into all the mapping from color to string and change the color code in the semantic map but it sounds like an inconvinient way to solve the problem. Would you mind any way to fix this ?  Thank !

**Comment by roozbehm** [comment]
We will fix this, but it is not a high priority now.

**Comment by phongnhhn92** [comment]
Thanks @roozbehm !
Actually, I am willing to help.
I have already take a look at the some C# script file in the Asset folder to find the part of code that you guys have used to generate semantic masks. 
I found something about the function _addImageSynthesisImageForm_ at the file AgentManager.cs like this.
`for (int i = 0; i < this.thirdPartyCameras.Count; i++) {
			ThirdPartyCameraMetadata cMetadata = new ThirdPartyCameraMetadata();
			Camera camera = thirdPartyCameras.ToArray()[i];
			cMetadata.thirdPartyCameraId = i;
			cMetadata.position = camera.gameObject.transform.position;
			cMetadata.rotation = camera.gameObject.transform.eulerAngles;
			cameraMetadata[i] = cMetadata;
			ImageSynthesis imageSynthesis = camera.gameObject.GetComponentInChildren<ImageSynthesis> () as ImageSynthesis;
			addThirdPartyCameraImageForm (form, camera);
			addImageSynthesisImageForm(form, imageSynthesis, this.renderDepthImage, "_depth", "image_thirdParty_depth");
			addImageSynthesisImageForm(form, imageSynthesis, this.renderNormalsImage, "_normals", "image_thirdParty_normals");
			addImageSynthesisImageForm(form, imageSynthesis, this.renderObjectImage, "_id", "image_thirdParty_image_ids");
			addImageSynthesisImageForm(form, imageSynthesis, this.renderClassImage, "_class", "image_thirdParty_classes");
		}`
I guess you guys are trying to generate all kinds of images (semantic, instance,depth map) on 3rd Part camera with this function. However, where is the part of the code that you select colors for the semantic maps ?

**Comment by phongnhhn92** [comment]
Is there any updates on this, guys ?
I have tried to generate data for the task of semantic segmentation but I found out that same objects class are having different color code.
file:///home/phong/data/ai2thor_data_creation-reconstruction/img/1/seg/FloorPlan1_s10_r19_rot310_hor-16_kp180.png
Take a look at this semantic example in the Scene 1 (kitchen), there are 3 ceiling lamps and I think they are in the same class but somehow, they have 3 different color codes when I check them in the `color_to_object_id` from the metadata. I whether this is a bug or not ?

**Comment by winthos** [comment]
So as an update, the main problem with objects in scenes like those lamps are that not all objects that are rendered are tagged the same way. Interactable objects in the scene are all classified as SimObjects, and each SimObject has an associated Type. Non sim objects we call Structure objects, these have collision and are rendered but do not report back metadata and can't be interacted with like Sim Objects. Some examples of Structure objects are walls, ceiling, floor, and most light fixtures.

Currently, the instance segmentation image will show different colors for every single object rendered. The class segmentation image uses the SimObject's Type to make sure all objects of the same type are colored the same (ie: all Microwaves are the same color across scenes).

This unified class segmentation color does not currently support Structure typed object like those light fixtures, as those derive their color based entirely on the string name of the object in the Unity Editor's heirarchy. Because structure objects don't currently have standardized names, you get results like you saw of the Wall in different scenes being different colors. 

We can fix this by basically having a tag for structural objects, and a fix for this is in the works. We should be able to release this fix relatively soon, but this requires hand tagging every rendered object in the scene that isn't already tagged by being a Sim Object.

**Comment by phongnhhn92** [comment]
@winthos Thanks for your response, I am looking for that fix soon. I am working on scene understand task so having this fix will help me a lot.

**Comment by phongnhhn92** [comment]
@winthos Hello, is there any updates on this one ?
If you guys are busy with other stuff then I can help. If you can show me a quick example how to change the object type then I will do it by myself and upload scene files later on this Github.
Thanks !

**Comment by winthos** [comment]
This has been fixed in our current working build, but we have a few things left to finish before we can push the update to the live master. I ended up having to come up with an alternate way of tagging objects to allow this to work, so it's not as simple as finding the object in the scene and using the Unity tags to sort objects into types. 

If you want to start using the updated segmentation immediately, you can switch to the WIPFeatures branch, but be aware that branch is in a state of flux and could break at anytime since we are actively working on it. These changes will be in our next live master release.

**Comment by phongnhhn92** [comment]
Thanks in that case, I will wait for the next release.

**Comment by mattdeitke** [comment]
Closing as this issue appears to be fixed :)
